blurple.py

```diff
@@ -1,58 +1,108 @@
-
-
-
-# https://www.python.org/downloads/ v3.6.5
-
-from config import *
-import discord # discord.py rewrite
-# pip install -U git+https://github.com/Rapptz/discord.py@rewrite#egg=discord.py[voice]
-from discord.ext import commands
-from discord.ext.commands.cooldowns import BucketType
-import asyncio
-from PIL import Image, ImageEnhance, ImageSequence
-import PIL
-#pip install Pillow
-from io import BytesIO
-import io
+import traceback
 import datetime
 import aiohttp
 import copy
-import sys
-import time
-from resizeimage import resizeimage
-#pip install python-resize-image
 import math
+import time
+import io
+import os
+import re
+
+import numpy as np
+import discord
+
+from PIL import Image, ImageEnhance, ImageSequence
+
+from discord.ext.commands.cooldowns import BucketType
+from discord.ext import commands
+
+try:
+    from config import TOKEN, BOT_PREFIX
+except ModuleNotFoundError:
+    # Probably in a docker container, just read environment
+    TOKEN = os.environ.get('TOKEN')
+    BOT_PREFIX = os.environ.get('BOT_PREFIX')
 
 description = '''Blurple Bot'''
 bot = commands.Bot(command_prefix=BOT_PREFIX, description=description)
 
 rocked = 204778476102877187
 
-blurple = (114, 137, 218)
-bluplehex = 0x7289da
-darkblurple = (78, 93, 148)
-white = (255, 255, 255)
+BLURPLE = (114, 137, 218, 255)
+BLURPLE_HEX = 0x7289da
+DARK_BLURPLE = (78, 93, 148, 255)
+WHITE = (255, 255, 255, 255)
+
+PIXEL_COUNT_LIMIT = 3840 * 2160
+MAX_PIXEL_COUNT = 1280 * 720
+MAX_FILE_SIZE = 8 * 1024 * 1024 * 16  # 16M
+
+COLOUR_BUFFER = 20
+
+MENTION_RE = re.compile(r'<@!?([0-9]+)>')
 
 bot.remove_command('help')
 
-allowedusers = {204778476102877187, 226595531844091904, 191602259904167936} #put your user id here, and it will allow you to use the 'hidden' commands (and shutdown command)
-approved_channels = {418987056111550464, 436300339273269278}
+allowed_users_list = set([int(i) for i in os.environ.get('ALLOWED_USERS').split(',')])
+approved_channels = set([int(i) for i in os.environ.get('APPROVED_CHANNELS').split(',')])
+
+
+class ImageStats:
+    __slots__ = ('dark', 'blurple', 'white', 'pixels')
+
+    def __init__(self, dark, blurple, white, pixels):
+        self.dark = dark
+        self.blurple = blurple
+        self.white = white
+        self.pixels = pixels
+
+    @property
+    def total(self):
+        return self.dark + self.blurple + self.white
+
+    def percentage(self, value):
+        return round(value / self.pixels * 100, 2)
+
+    @classmethod
+    def from_image(cls, img):
+        arr = np.asarray(img).copy()
+
+        dark = np.all(np.abs(arr - DARK_BLURPLE) < COLOUR_BUFFER, axis=2).sum()
+        blurple = np.all(np.abs(arr - BLURPLE) < COLOUR_BUFFER, axis=2).sum()
+        white = np.all(np.abs(arr - WHITE) < COLOUR_BUFFER, axis=2).sum()
+        pixels = img.size[0] * img.size[1]
+
+        mask = np.logical_and(np.abs(arr - DARK_BLURPLE) >= COLOUR_BUFFER,

```

Comment:
Somewhere in this part regarding the mask, it causes the blurple check to output white pixels as black aka non-blurple.
I don't know enough python to fix it, but I know enough to trace it down to this part

Reply:
It's actually just below line 80 😉. Working on finding what's causing it.

Reply:
Awesome